File: cpp/1930.cpp
class Solution {
public:
    int countPalindromicSubsequence(string s) {
        vector<vector<bool>> seen(26, vector<bool>(26, false));
        vector<int> first(26, -1);
        int count = 0;
        
        for (int i = 0; i < s.length() - 2; i++) {
            if(first[s[i] - 'a'] != -1){
                continue;
            }
            first[s[i] - 'a'] = i;
            char c = s[i];
            int firstIdx = i;
            int lastIdx = -1;
            for (int j = s.length() - 1; j > i + 1; j--) {
                if (s[j] == c) {
                    lastIdx = j;
                    break;
                }
            }
        
            if (lastIdx == -1) continue;
            
            for (int k = firstIdx + 1; k < lastIdx; k++) {
                int midChar = s[k] - 'a';
                if (!seen[c - 'a'][midChar]) {
                    seen[c - 'a'][midChar] = true;
                    count++;
                }
            }
        }
            
        return count;
    }
};

File: python/1930.py
class Solution:
    def countPalindromicSubsequence(self, s: str) -> int:
        seen = [set() for _ in range(26)]
        processed = [False] * 26
        count = 0
        
        for i in range(len(s) - 2):
            outer_char = s[i]
            outer_idx = ord(outer_char) - ord('a')
            
            if processed[outer_idx]:
                continue
            
            processed[outer_idx] = True
            first_idx = i
            
            last_idx = s.rfind(outer_char, first_idx + 2)
            
            if last_idx == -1:
                continue
            
            middle_chars = set(s[first_idx + 1:last_idx])
            count += len(middle_chars)
            seen[outer_idx] = middle_chars
        
        return count

